src/node_report.h

```diff
@@ -0,0 +1,190 @@
+#ifndef SRC_NODE_REPORT_H_
+#define SRC_NODE_REPORT_H_
+
+#include "v8.h"
+#include <uv.h>
+#include <node.h>
+#include <node_buffer.h>
+#include <algorithm>
+#include <cstring>
+#include <climits>
+#include <cstdlib>
+#include <utility>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <iostream>
+#include <sstream>
+#include <iomanip>
+
+#ifdef _WIN32
+#include <time.h>
+#else
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/time.h>
+#endif
+
+namespace nodereport {
+
+// enum DumpEvent {kException, kFatalError, kSignal_JS, kSignal_UV,
+// kJavaScript};
+
+#ifdef _WIN32
+typedef SYSTEMTIME TIME_TYPE;
+typedef DWORD PID_TYPE;
+#define PATHSEP "\\"
+#else  // UNIX, OSX
+typedef struct tm TIME_TYPE;
+typedef pid_t PID_TYPE;
+#define PATHSEP "/"
+#endif
+
+typedef struct write_data {
+  std::ostream* out;
+  int* intend;
+} write_data;
+
+#define UNKNOWN_NODEVERSION_STRING "Unable to determine Node.js version\n"
+
+void InitializeNodeReport(v8::Isolate* isolate, node::Environment* env);
+
+// Function declarations - functions in src/node_report.cc
+void TriggerNodeReport(v8::Isolate* isolate, node::Environment* env,
+                       const char* message, const char* location,
+                       char* name, v8::MaybeLocal<v8::Value> error);
+void GetNodeReport(v8::Isolate* isolate,
+                   node::Environment* env,
+                   const char* message,
+                   const char* location,
+                   v8::MaybeLocal<v8::Value> error,
+                   std::ostream& out);
+
+// Function declarations - utility functions in src/utilities.cc
+void ReportEndpoints(uv_handle_t* h, std::ostringstream& out);
+void WalkHandle(uv_handle_t* h, void* arg);
+void WriteInteger(std::ostream& out, size_t value);
+#if defined(HAVE_OPENSSL)
+std::string GetOpenSSLVersion();
+#endif  // HAVE_OPENSSL
+
+
+// Function declarations - export functions in src/node_report_module.cc
+void TriggerReport(const v8::FunctionCallbackInfo<v8::Value>& info);
+void GetReport(const v8::FunctionCallbackInfo<v8::Value>& info);
+
+
+// Node.js boot time - defined in src/node_report.cc
+extern time_t boot_time;
+
+// JSON compiler definitions.
+#define ADVANCE(jstr, intend)                                                  \
+  do {                                                                         \
+    for (int i = 0; i < intend; i++) jstr << " ";                              \
+  } while (0)
```

Comment:
I think all of these macros could be inline functions, right?

Reply:
@addleax - I attempted an inline version. The difficulty I am seeing is with parameter passing:

with macro, I was able to pass any primitive typed values and std::Strings without really caring about those and the osrteam was able to accommodate those. Now with inline functions, I am needing to worry about passing and receiving indifferent types:

```c++
template <typename T> inline void json_keyvaluenormal(std::ostream& out, int indent, T key, T value)
```

with this as the inline function calls of these types are not compiling for obvious reasons:
```c++
json_keyvaluenormal(std::ostream&, int, const char*, char*);
json_keyvaluenormal(std::ostream&, int, char*, const char*);
json_keyvaluenormal(std::ostream&, int, const char*, std::string&);
json_keyvaluenormal(std::ostream&, int, char*, std::string&);
json_keyvaluenormal(std::ostream&, int, int, char*); ...
```
we need to be precise, for apparently no reason. Are you ok to retain those as macros?

Reply:
@gireeshpunathil In the case of `json_keyvaluenormal`, you could use templates for both parameters, right?

> Are you ok to retain those as macros?

You can keep them but I’d probably do the conversion myself afterwards – which is totally fine, if you prefer that

Reply:
@addaleax - thanks, done. I was not following the `template` feature properly, and hence the confusion. With:
```c++
template <typename T, typename U>
inline void json_keyvaluenormal(std::ostream& out, int indent, T key, U value)
```
every existing calls went through smooth. thanks!